### PR TITLE
Merge preprocess_img and rescale_img

### LINKED ISSUE #619: Redundancy between ``deepinv.utils.plotting.rescale_img`` and ``deepinv.utils.plotting.preprocess_img``
Both functions have very similar roles and are used interchangeably in the plot functions.
I think it would be better to merge them. For reference, here's the code of both functions:

```python
def preprocess_img(im, rescale_mode="min_max"):
    with torch.no_grad():
        if im.shape[1] == 2:  # for complex images
            pimg = im.pow(2).sum(dim=1, keepdim=True).sqrt().type(torch.float32)
        elif im.shape[1] > 3:
            pimg = im.type(torch.float32)
        else:
            if torch.is_complex(im):
                pimg = im.abs().type(torch.float32)
            else:
                pimg = im.type(torch.float32)

        pimg = rescale_img(pimg, rescale_mode=rescale_mode)
    return pimg 
```
---
```
def rescale_img(im, rescale_mode="min_max"):
    img = im.clone()
    if rescale_mode == "min_max":
        shape = img.shape
        img = img.reshape(shape[0], -1)
        mini = img.min(1)[0]
        maxi = img.max(1)[0]
        idx = mini < maxi
        mini = mini[idx].unsqueeze(1)
        maxi = maxi[idx].unsqueeze(1)
        img[idx, :] = (img[idx, :] - mini) / (maxi - mini)
        img = img.reshape(shape)
    elif rescale_mode == "clip":
        img = img.clamp(min=0.0, max=1.0)
    else:
        raise ValueError("rescale_mode has to be either 'min_max' or 'clip'.")
    return img
```

### PR DESCRIPTION
**Pull request by jscanvic**
Fix #619

### CODE REVIEW

File: deepinv/utils/plotting.py [hunk 1]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+
+    return im
 
-        pimg = rescale_img(pimg, rescale_mode=rescale_mode)
-    return pimg
+
+@torch.no_grad
+def preprocess_img(im: torch.Tensor, rescale_mode: str = "min_max"):
+    r"""
+    Preprocesses an image tensor for plotting.
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: Please can this have a more informative docstring? I.e. what is preprocessing if it is not just rescaling?

File: deepinv/utils/plotting.py [hunk 2]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+
+    # NOTE: Rescaling a constant image between zero and one is ill-defined.
+    # Indeed, no matter the new scale, the image won't have a minimum value
+    # equal to zero and a maximum value equal to one. In this case, we
+    # choose to rescale the image so that its new value is the number in
+    # [0, 1] that is closest to its original value. This amounts to
+    # clamping it between zero and one.
+    if rescale_mode == "min_max":
+        # Compute batch-wise minimum and maximum values
+        # NOTE: Expect a batched input of shape (B, *)
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: can you add instead this NOTE one liner to the param description of im for all 3 functions

File: deepinv/utils/plotting.py [hunk 3]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+            # and imaginary part of a complex-valued image.
+            im = torch.linalg.vector_norm(im, ord=2, dim=channels_dim, keepdim=True)
+
+        # Cast image values to float32 numbers
+        im = im.type(torch.float32)
+
+    min_val = 0.0
+    max_val = 1.0
+
+    # NOTE: Rescaling a constant image between zero and one is ill-defined.
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: Nice, I like this solution. I think this comment is a bit verbose - can you just incorporate this comment into the two comments below starting "Rescale non-constant ..." and "Clamp constant..."

File: deepinv/utils/plotting.py [hunk 4]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+    # Indeed, no matter the new scale, the image won't have a minimum value
+    # equal to zero and a maximum value equal to one. In this case, we
+    # choose to rescale the image so that its new value is the number in
+    # [0, 1] that is closest to its original value. This amounts to
+    # clamping it between zero and one.
+    if rescale_mode == "min_max":
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: Do you think we should test this small behaviour?

File: deepinv/utils/plotting.py [hunk 5]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+    if not rescale_only:
+        # Apply the modulus function on each entry if the input is complex-valued
+        # or if it has 2 channels which are assume to represent the real and
+        # imaginary parts of a complex-valued image.
+        # NOTE: Expect a batched input
+        channels_dim = 1
+        n_channels = im.shape[channels_dim]
+        is_complex = torch.is_complex(im) or n_channels == 2
+        if is_complex:
+            # NOTE: This implementation is meant to work for complex-valued images
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: Remove repeated comment

File: deepinv/utils/plotting.py [hunk 6]
```diff
@@ -115,27 +115,79 @@ def prepare_images(x=None, y=None, x_net=None, x_nl=None, rescale_mode="min_max"
...
+            # and for real-valued images with two channels, understood as the real
+            # and imaginary part of a complex-valued image.
+            im = torch.linalg.vector_norm(im, ord=2, dim=channels_dim, keepdim=True)
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py]: We already have this function `deepinv.loss.metric.functional.complex_abs` https://github.com/deepinv/deepinv/blob/7116f703c115bad391ab83f6b9fa6f5ccc85c75d/deepinv/loss/metric/functional.py#L39 do you think you could unify this line and this function?

**Review by Andrewwango — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by jscanvic** [comment]
Thanks for the suggestions Andrew - I made the changes!

File: deepinv/utils/plotting.py [hunk 7]
```diff
@@ -200,15 +161,18 @@ def numpy2uint(img):
...
 
     :param torch.Tensor im: the image to rescale.
     :param str rescale_mode: the rescale mode, either 'min_max' or 'clip'.
     :return: the rescaled image.
     """
-    return _preprocess_img(im, rescale_only=True, rescale_mode=rescale_mode)
+    warn(
+        "The function `deepinv.utils.plotting.rescale_img` is deprecated and will be removed in a future version. Use `deepinv.utils.normalize_signal` instead."
+    )
+    return normalize_signal(im, mode=rescale_mode)
```
Comment (Andrewwango) [inline deepinv/utils/plotting.py:177]: there are a number of other things that use rescale_img. Please can you switch them to normalize_signal?

Comment (jscanvic) [reply deepinv/utils/plotting.py:177]: I swear I knew I was forgetting something! Thanks for catching that!

**Comment by jscanvic** [comment]
Should be good now!

File: deepinv/utils/signal.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        maximum_intensity = maximum_intensity.view(*shape)
+
+        # Rescale the non-constant batched signals between zero and one
+        inp[indices] -= minimum_intensity[indices]
+        inp[indices] /= maximum_intensity[indices] - minimum_intensity[indices]
+
+        # The indices corresponding to the constant batched signals
+        indices = torch.logical_not(indices)
+
+        # Clamp constant batched signals between zero and one
```
Comment (mh-nguyen712) [inline deepinv/utils/signal.py:38]: Should we devide the constant batch by the maximum_intensity before clip?

Comment (mh-nguyen712) [reply deepinv/utils/signal.py:38]: Sorry I changes nothing since the images are constant

**Review by mh-nguyen712 — APPROVED** [review]

**Comment by jscanvic** [comment]
Thanks for the review Hai!

**Review by Andrewwango — APPROVED** [review]
